lib/syslog/pool/pool.go

```diff
@@ -0,0 +1,92 @@
+package pool
+
+import (
+	"io"
+)
+
+// A LimitedConnPool is a connection pool, with the property that
+// the total number of live connections is limited to the size
+// parameter passed to NewLimited. In order to make this guarantee,
+// we assume that all new connections are introduced by calls to the
+// Get method, and return to the pool by the Put method (whether
+// or not the connection is dead)
+type LimitedConnPool struct {
+	conns  chan io.WriteCloser // Available connections in the pool
+	live   chan struct{}       // Keep a count of living connections (in our hands, or the client)
+	signal chan struct{}       // Used to wake up the connection producer
+	err    chan error          // Send dial errors back to the client
```

Comment:
Is there an assumption that the `conns` and `live` content of the channels will be synchronized? If so, the code doesn't yet seem to take any care to keep them synchronized via a mutex or the like.

Reply:
These will not be in sync, and they needn't be. `live` acts as a counter of all connections which have been created, and not returned marked "dead", i.e. living connections tracked by the pool. `conns` simply holds available connections to be passed to the client. `conns` provides the blocking mechanism if there are no available connections (`len(live) == size`), and `live` provides the rendezvous point in the connection producer goroutine. `signal` lets the producer avoid a busy wait, and fires whenever `live` is decremented.